File: python/108_Convert_Sorted_Array_to_Binary_Search_Tree.py
#!/usr/bin/env python3

# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        if nums == None or len(nums) == 0:
            return None

        root = TreeNode(0)
        self.helper(nums, 0, len(nums), root)
        return root

    def helper(self, nums, start, end, root):
        if end - start == 1:
            root.val = nums[start]
            return

        mid = (start + end) // 2

        root.val = nums[mid]

        if mid - start > 0:
            root.left = TreeNode(0)
            self.helper(nums, start, mid, root.left)
        if end - 1 - mid > 0:
            root.right = TreeNode(0)
            self.helper(nums, mid + 1, end, root.right)

File: java/src/brand_new/_108_Convert_Sorted_Array_to_Binary_Search_Tree.java
package brand_new;

public class _108_Convert_Sorted_Array_to_Binary_Search_Tree {
	public TreeNode sortedArrayToBST(int[] nums) {
		return sortedArrayToBST(nums, 0, nums.length);
	}

	public TreeNode sortedArrayToBST(int[] nums, int i, int j) {
		if (i >= j)
			return null;
		if (j - i == 1)
			return new TreeNode(nums[i]);

		int mid = (i + j) / 2;
		TreeNode head = new TreeNode(nums[mid]);
		head.left = sortedArrayToBST(nums, i, mid);
		head.right = sortedArrayToBST(nums, mid + 1, j);
		return head;
	}
}
